### DISCUSSION TITLE
Electron dynamics with custom external fields

### DISCUSSION THREAD

**Discussion by samusake55** [dftbplus/dftbplus#1175] [Q&A] [answered]
Hello you all,

for my research i want to do an electron-dynamics simulation using the dftb+ code. 
In the input file one can select different types of laser envelopes (limited to Constant, Gaussian, Sin2). 
Is there a way to specify different envelopes, e.g. Trapezoidal, without modifying the source code?

I found the option "fromFile" in this part of the source-code ([https://github.com/dftbplus/dftbplus/blob/release/src/dftbp/timedep/timeprop.F90]), but was unable to figure out how to apply it.

Kind regards and thanks to everybody who contributed to this great software.

**Comment by aradi** [comment]
The option is not really documented, I am sorry for that. According to the code which is relevant here, you would need
```
EnvelopeShape = FromFile {}
```
and a ``laser.dat`` file containing nSteps + 1 lines (1st lines corresponding to 0 time), where every line has the form
```
time E_x E_y E_z
```
where the field component are specified in eV/ang. The time is actually irrelevant as I can judge based on

https://github.com/dftbplus/dftbplus/blob/2bc8d32aae6434c9522b7df3617ae8dd207614f8/src/dftbp/timedep/timeprop.F90#L1348

@fbonafe @charlymedrano7 can you confirm on this? Also, is it safe to use this non-documented option?

**Reply by samusake55** [reply]
Thank you for your reply!

I played a little bit around with what you said and found the following.
The laser.dat you specified looks exactly like the one generated by dftb+. Therefore I performed a simulation with a 4000nm sin^2 shaped pulse and  then used that generated laser.dat file to try and get similar results.

Prompt1:
```
ElectronDynamics = {
   Steps = 275580
   TimeStep [au] = 0.02
   Perturbation = Laser {
     PolarizationDirection = 1 0 0
     LaserEnergy [au] = 0.0114  #2000NM = 0.0228
   }
   FieldStrength [au] = 0.02
   EnvelopeShape = Sin2 {
      Time1 [au] = 5511.566059      #one period 2000nm =2pi/omega=275.57830295 au
   }
}
```
yields an expected response.



Prompt2:
```
ElectronDynamics = {
   Steps = 275580
   TimeStep [au] = 0.02
   Perturbation = Laser {
     PolarizationDirection = 1 0 0
     LaserEnergy [au] = 0.0114  #2000NM = 0.0228
   }
   FieldStrength [au] = 0.02
   EnvelopeShape = fromFile {}
}
```
throws the error 
```
At line 1348 of file /home/conda/feedstock_root/build_artifacts/dftbplus_1672913244616/work/src/dftbp/timedep/timeprop.F90
Fortran runtime error: Cannot read from file opened for WRITE
```
which makes sense, because I guess if I tell the program to `Pertubation = Laser{...}`, it will try to write a new laser.dat file. (also the envelope has the laser already included)



Prompt3:

```
ElectronDynamics = {
   Steps = 275580
   TimeStep [au] = 0.02
   EnvelopeShape = fromFile {}
}
```

works but the resulting dipole moments are absolute nonsense and not the same as generated by prompt1.


I will try to modify the code later on by myself, if i haven't made any obvious mistakes and this part of the code is just too experimental. It is not so urgent.

Thanks again for helping ^-^

**Comment by aradi** [comment, marked as answer]
OK, that's indeed a bug. The good news is, I have a big PR (#1151), which changes the file handling everywhere in the code. As a side effect, I fixed the `laser.dat` problem. So, you should be able to use `laser.dat` when specifying `FromFile {}` as `Envelope`.  If you use a `laser.dat` created by DFTB+, you would have to remove the first line (the comment), though. Once the PR had been merged, the feature should be available on main (and become part of the next release).

**Comment by samusake55** [comment]
Thank you, that's amazing!

**Comment by aradi** [comment]
FYI: #1151 had been merged, so it should work with the code on main.
